compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1481,6 +1481,7 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
 
                 let mut err;
                 let mut unsized_return = false;
+                let mut ret_vec: Vec<hir::HirId> = Vec::new(); // collect return expression, record HirId to save memory
```

Comment:
It's fine, `hir::Expr` should be cheap enough for this, and we are already emitting the error, the allocation will have to happen either way, we wouldn't be pessimizing the happy path.